File: BOJ/1148 단어 만들기.java
// 2022-09-28
// 1148 단어 만들기
// https://www.acmicpc.net/problem/1148
/*
    문자열, 구현 (112,812KB, 556ms)
    시간 복잡도: ?
    풀이)
    - 입력되는 단어 속 문자의 개수를 저장할 words[][]. (word[i]['A'] = k일 때, i번 단어 안에 문자 'A'는 k개 존재)
    - 입력된 퍼즐 속 문자의 개수를 저장할 m[].(m[c] = k일 때, 퍼즐에 문자 c는 k개 존재)
    - 단어를 완성할 때 사용한 문자의 횟수를 저장할 cnt[]. (cnt[c] = k일 때, 문자 c는 단어 완성에 k번 사용)
    - m을 이용하여 현재 문자 word를 만들 수 있는지 확인. (words[][]에 저장된 값을 사용)
    만약 퍼즐에 존재하는 문자의 개수보다 단어에 존재하는 문자의 개수가 많을 경우, 즉 m[c] < word[c]일 경우 만들 수 없음.
    만약 단어를 만들 수 있을 경우, 'A'부터 'Z'까지 확인.
     - 단어 속 문자가 존재할 경우, 즉 word[c]가 0이 아닐 경우 cnt[c]++.
    - 퍼즐 속 문자 c를 이용하여 cnt[c]를 확인하여 최소 개수 mn과 최대 개수 mx를 구함.
    - 최소 개수 mn을 이용한 문자들 s1과 최대 개수 mx를 이용한 문자들 s2를 구함.
    이때, 퍼즐에 존재해야 하므로 m[c]가 0보다 커야 한다.
    etc.)
    - String으로 words를 선언하면 MLE.
    - 제출한 두 코드의 차이는 2차원 배열 선언 시 []['Z']와 [][26] 뿐이지만 메모리 차이는 두 배. why?
 */

import java.util.*;
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class Main {
    static final int INF = 0x3f3f3f3f;
    static final int MAX = 200003; // max words.size()

    static int[][] words = new int[MAX]['Z' + 3]; // word[i]['A'] = k일 때, i번 단어 안에 문자 'A'는 k개 존재

    static boolean isValid(int[] m, int[] word) // 만들 수 있는 단어인가?
    {
        for (char c = 'A'; c <= 'Z'; c++)
            if (m[c] < word[c])
                return false;

        return true;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int sz = 0;
        for (;; sz++) /* input word */ {
            String word = br.readLine();
            if (word.equals("-"))
                break;
            for (char c : word.toCharArray())
                words[sz][c]++;
        }

        while (true) /* input puzzle */ {
            String puzzle = br.readLine();
            if (puzzle.equals("#"))
                break;

            int[] m = new int['Z' + 3]; // 퍼즐의 문자의 개수를 저장 (m[c] = k일 때, 퍼즐에 문자 c는 k개 존재)
            for (char c : puzzle.toCharArray())
                m[c]++;

            int[] cnt = new int['Z' + 3]; // 단어를 완성할 때 사용한 문자의 횟수를 저장 (cnt[c] = k일 때, 문자 c는 단어 완성에 k번 사용)
            for (int i = 0; i < sz; i++) {
                int[] word = words[i];

                if (!isValid(m, word))
                    continue;

                for (char c = 'A'; c <= 'Z'; c++)
                    if (word[c] > 0)
                        cnt[c]++;
            }

            int mn = INF, mx = 0;
            for (char c : puzzle.toCharArray()) {
                mn = Math.min(mn, cnt[c]);
                mx = Math.max(mx, cnt[c]);
            }

            String s1 = "", s2 = "";
            for (char c = 'A'; c <= 'Z'; c++) {
                if (m[c] > 0 && cnt[c] == mn)
                    s1 += c;
                if (m[c] > 0 && cnt[c] == mx)
                    s2 += c;
            }

            System.out.printf("%s %d %s %d\n", s1, mn, s2, mx); // mn == mx에 주의
        }
    }
}

File: BOJ/1148 단어 만들기.cpp
// 2022-09-28
// 1148 단어 만들기
// https://www.acmicpc.net/problem/1148
/*
    문자열, 구현 (14,444KB, 92ms)
    시간 복잡도: ?
    풀이)
    - 입력된 단어들을 저장할 words, 입력된 퍼즐판을 저장할 puzzle.
    - puzzle에 존재하는 문자의 개수를 저장할 m[].  (m[c] = k일 때, 퍼즐에 문자 c는 k개 존재)
    - words에 저장된 단어 word를 완성할 때 사용한 문자의 사용 횟수를 저장할 cnt[]. (cnt[c] = k일 때, 문자 c는 단어 완성에 k번 사용)
    - word에 존재하는 문자의 개수를 저장할 ch[].
    - m에 저장된 값을 이용하여 word가 만들 수 있는 문자인지 확인.
    만약 만들 수 있다면, 만들 때 사용된 문자를 cnt에 저장.
    이때, 중복된 문자는 한 번만 사용한 것으로 간주해야 함.
    - 최소 개수를 만들 수 있는 문자들을 저장할 문자열 s1, 최대 개수를 만들 수 있는 문자들을 저장할 문자열 s2.
    - cnt의 최솟값 mn과 최댓값 mx를 구함.
    이후, puzzle의 각 문자 c를 cnt에서 확인.
    만약 cnt[c] == mn일 경우, s1에 c를 추가.
    만약 cnt[c] == mx일 경우, s2에 문자를 추가.
    동일한 문자가 존재할 수 있으므로 값을 초기화하여 중복을 방지해야 한다.
    - 저장된 s1과 mn, s2와 mx를 출력한다.
    단, mn == mx일 경우 s2 대신 s1을 출력한다. (중복 방지로 인해 mn == mx일 때의 s2는 값이 저장되지 않음. s1으로 가기 때문)
    etc.)
    - 엄청 헷갈린 문제지만 다행히 원트
 */

#include <iostream>
#include <vector>
#include <algorithm> // sort

using namespace std;
#define FAST_IO ios::sync_with_stdio(false), cin.tie(NULL), cout.tie(NULL);
#define endl "\n"
#define INF 0x3f3f3f3f

vector<string> words; // words.size() ≤ 200,000

int main()
{
    FAST_IO;

    { // input word
        string s;
        while (cin >> s)
        {
            if (s == "-")
                break;
            words.push_back(s);
        }
    }
    { // input puzzle
        string puzzle;
        while (cin >> puzzle)
        {
            if (puzzle == "#")
                break;

            int m['Z' + 3]{}; // 퍼즐의 문자의 개수를 저장 (m[c] = k일 때, 퍼즐에 문자 c는 k개 존재)
            for (const char &c : puzzle)
                m[c]++;

            int cnt['Z' + 3]{}; // 단어를 완성할 때 사용한 문자의 사용 횟수를 저장 (cnt[c] = k일 때, 문자 c는 단어 완성에 k번 사용)
            for (const string &word : words)
            {
                bool valid = true; // 만들 수 있는 단어인가?
                int ch['Z' + 3]{}; // 문자의 개수를 저장
                for (const char &c : word)
                    if (m[c] < ++ch[c])
                        valid = false;

                if (!valid)
                    continue;

                for (const char &c : puzzle)
                {
                    if (ch[c] == -1) // 중복 제외
                        continue;
                    if (ch[c] != 0) // 사용된 문자인가
                        cnt[c]++;
                    ch[c] = -1;
                }
            }

            int mn = INF, mx = 0;
            for (const char &c : puzzle)
                mn = min(mn, cnt[c]), mx = max(mx, cnt[c]);

            string s1, s2;
            for (const char &c : puzzle)
            {
                if (cnt[c] == mn)
                    s1.push_back(c), cnt[c] = -1;
                if (cnt[c] == mx)
                    s2.push_back(c), cnt[c] = -1;
            }

            sort(s1.begin(), s1.end()), sort(s2.begin(), s2.end());
            cout << s1 << ' ' << mn << ' ' << (mn == mx ? s1 : s2) << ' ' << mx << endl; // mn == mx에 주의
        }
    }
}